irohad/ametsuchi/impl/postgres_command_executor.cpp

```diff
@@ -0,0 +1,770 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ametsuchi/impl/postgres_command_executor.hpp"
+
+#include <boost/format.hpp>
+
+#include "ametsuchi/impl/postgres_wsv_common.hpp"
+
+namespace iroha {
+  namespace ametsuchi {
+
+    expected::Error<CommandError> makeCommandError(
+        const std::string &error_message,
+        const std::string &command_name) noexcept {
+      return expected::makeError(CommandError{command_name, error_message});
+    }
+
+    /**
+     * Transforms soci statement to CommandResult,
+     * which will have error message generated by error_generator
+     * appended to error received from given result
+     * @param result which can be received by calling execute_
+     * @param error_generator function which must generate error message
+     * to be used as a return error.
+     * Function is passed instead of string to avoid overhead of string
+     * construction in successful case.
+     * @return CommandResult with combined error message
+     * in case of result contains error
+     */
+    template <typename Function>
+    CommandResult makeCommandResult(
+        soci::statement &st,
+        std::string command_name,
+        Function &&error_generator) noexcept {
+      st.define_and_bind();
+      try {
+        st.execute(true);
+      } catch (std::exception &e) {
+        return makeCommandError(error_generator() + "\n" + e.what(),
+                                command_name);
+      }
+      return {};
+    }
+
+    /**
+     * Transforms soci statement to CommandResult,
+     * which will have error message generated exception
+     * Assums that statement query returns 0 in case of success or error code
+     * @param result which can be received by calling execute_
+     * @param error_generator functions which must generate error message
+     * to be used as a return error.
+     * Functions are passed instead of string to avoid overhead of string
+     * construction in successful case.
+     * @return CommandResult with combined error message
+     * in case of result contains error
+     */
+    CommandResult makeCommandResultByReturnedValue(
+        soci::statement &st,
+        std::string command_name,
+        std::vector<std::function<std::string()>> &error_generator) noexcept {
+      uint32_t result;
+      st.exchange(soci::into(result));
+      st.define_and_bind();
+      try {
+        st.execute(true);
+        for (size_t i = 0; i < error_generator.size(); i++) {
```

Comment:
Please remove this loop (use result + 1 as index)

Reply:
Facepalm